include/xpedite/common/LeasedPtr.H

```diff
@@ -0,0 +1,99 @@
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// LeasedPtr - a SPSC wait free data structure for sharing a resource between two threads.
+//
+// It has a clear ownership semantics, allowing the consumer to lease the pointer from the producer.
+//
+// Once the consumer has successfully borrowed the resource, the producer will know not to deallocate
+// the resource until the consumer has returned the lease.
+//
+// If the consumer takes too long to return the resource, the producer can release the ownership.
+// The LeasedPtr will ensure the consumer knows it now owns the resource and is responsible for deallocation.
+//
+// Author: Marcin Dlugajczyk, Morgan Stanley
+//
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+#include <atomic>
+
+namespace xpedite { namespace common {
+
+  namespace {
+    static constexpr char leased{};
+  }
+
+  template<typename T>
+  class LeasedPtr
```

Comment:
should this be `LeaseablePtr` ?

Reply:
i don't mind either way, but would `LeasedPtr` be more consistent with `std::shared_ptr`? we don't have `std::sharable_ptr'